File: alp-portable/HYDRA/hydra/core/context_engine.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass


@dataclass
class ContextSlice:
    source: str
    content: str
    tokens_estimate: int = 0


class ContextEngine:
    def __init__(self, max_tokens: int = 4096) -> None:
        self._store: dict[str, list[ContextSlice]] = {}
        self._max_tokens = max_tokens

    def add(self, key: str, slice: ContextSlice) -> None:
        self._store.setdefault(key, []).append(slice)

    def build_prompt(self, key: str, query: str) -> str:
        slices = self._store.get(key, [])
        selected: list[str] = []
        total = 0
        for s in slices:
            if total + s.tokens_estimate > self._max_tokens:
                break
            selected.append(s.content)
            total += s.tokens_estimate
        return "\n\n".join(selected + [query])

    def fingerprint(self, text: str) -> str:
        return hashlib.sha256(text.encode("utf-8")).hexdigest()[:16]

File: alp-portable/HYDRA/crates/hydra-core/src/context_engine.rs
use std::collections::HashMap;

pub struct ContextEngine {
    pub max_tokens: usize,
    pub store: HashMap<String, Vec<ContextSlice>>,
}

pub struct ContextSlice {
    pub source: String,
    pub content: String,
    pub tokens_estimate: usize,
}

impl ContextEngine {
    pub fn new(max_tokens: usize) -> Self {
        Self {
            max_tokens,
            store: HashMap::new(),
        }
    }

    pub fn add(&mut self, key: &str, slice: ContextSlice) {
        self.store.entry(key.into()).or_default().push(slice);
    }

    pub fn build_prompt(&self, key: &str, query: &str) -> String {
        let mut out = Vec::new();
        let mut total = 0usize;
        if let Some(slices) = self.store.get(key) {
            for s in slices {
                if total + s.tokens_estimate > self.max_tokens {
                    break;
                }
                total += s.tokens_estimate;
                out.push(s.content.clone());
            }
        }
        out.push(query.into());
        out.join("\n\n")
    }
}
